src/source-maps.cc

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "source-maps.h"
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+
+#include "json/json.h"
+
+#define INDEX_NONE static_cast<size_t>(-1)
+
+#define INVALID()       \
+  do {                  \
+    if (fatal) abort(); \
+    return false;       \
+  } while (0);
+
+bool SourceMap::Validate(bool fatal) const {
+  for (size_t i = 0; i < segment_groups.size(); ++i) {
+    const auto& group = segment_groups[i];
+    if (i > 0 && group.generated_line <= segment_groups[i - 1].generated_line) {
+      INVALID();
+    }
+    for (size_t j = 0; j < group.segments.size(); ++j) {
+      const auto& seg = group.segments[j];
+      const Segment* last_seg = nullptr;
+      if (j > 0) last_seg = &group.segments[j - 1];
+      if (seg.generated_col_delta == 0) INVALID();
+      if (!seg.has_source && seg.has_name) INVALID();
+      if (!seg.has_source) return true;
+      if (seg.source >= sources.size()) INVALID();
+      if (last_seg) {
+        if (seg.source_line_delta == 0 && seg.source_col_delta == 0) INVALID();
+        // FIXME: This has a limitation that if this seg has a source, the last
+        // one must.
+        if (last_seg->source_line + seg.source_line_delta != seg.source_line) {
+          INVALID();
+        }
+        if (last_seg->source_col + seg.source_col_delta != seg.source_col) {
+          INVALID();
+        }
+      }
+      if (!seg.has_name) return true;
+      if (seg.name >= names.size()) INVALID();
+    }
+  }
+  return true;
+}
+
+static int32_t cmpLocation(const SourceMapGenerator::SourceLocation& a,
```

Comment:
We're still pretty inconsistent, but I think this should eventually be `CmpLocation`